In [1]:
import graphlab

## selection and summary statistics

In [2]:
sales = graphlab.SFrame('home_data.gl/')
sales.show()

This non-commercial license of GraphLab Create for academic use is assigned to r38411@yahoo.com and will expire on August 31, 2019.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1537972562.log


Canvas is accessible via web browser at the URL: http://localhost:49260/index.html
Opening Canvas in default web browser.


In [ ]:
### find out which neighbor has the average price

In [3]:
sales.show(view='BoxWhisker Plot', x='zipcode', y='price')

Canvas is accessible via web browser at the URL: http://localhost:49260/index.html
Opening Canvas in default web browser.


In [4]:
import graphlab.aggregate as agg

In [5]:
graphlab.canvas.set_target('ipynb')

In [6]:
zipcode_avg = sales.groupby(key_columns='zipcode', operations={'mean_price':agg.MEAN('price')})
zipcode_avg = zipcode_avg.sort('mean_price',ascending=False)

In [7]:
zipcode_avg[0]

{'mean_price': 2160606.6, 'zipcode': '98039'}

# Question 1: The zipcode: 98039 has the highest average home price=2160606.6

In [ ]:
## Question 2: filter data

In [8]:
median_size = sales[( sales['sqft_living']>=2000) & (sales['sqft_living'] <=4000)]
print len(sales), len(median_size), len(median_size)/float(len(sales))

21613 9221 0.426641373248


# Question 2: The fraction of house size between 2000 and 40000 is 0.42664

In [ ]:
## Question 3: build a regression model

In [14]:
simple_feature = ['sqft_living']
my_feature = ['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'zipcode']
extra_feature = [
    'condition', 'grade', 'waterfront', 'view', 'sqft_above', 'sqft_basement',
    'yr_built', 'yr_renovated', 'lat', 'long', 'sqft_living15',
    'sqft_lot15',
]
advanced_feature = more_feature + extra_feature

In [15]:
train_data, test_data = sales.random_split(.8, seed=0)

In [17]:
my_feature_model = graphlab.linear_regression.create(train_data, target='price', features=my_feature,
                                                    validation_set=None)

Linear regression:

--------------------------------------------------------

Number of examples          : 17384

Number of features          : 6

Number of unpacked features : 6

Number of coefficients    : 115

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+--------------+--------------------+---------------+

| 1         | 2        | 1.031930     | 3763208.270523     | 181908.848367 |

+-----------+----------+--------------+--------------------+---------------+

SUCCESS: Optimal solution found.

In [18]:
advanced_feature_model = graphlab.linear_regression.create(train_data, target='price',
                                                          features=advanced_feature,
                                                          validation_set=None)

Linear regression:

--------------------------------------------------------

Number of examples          : 17384

Number of features          : 18

Number of unpacked features : 18

Number of coefficients    : 127

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+--------------------+---------------+

| Iteration | Passes   | Elapsed Time | Training-max_error | Training-rmse |

+-----------+----------+--------------+--------------------+---------------+

| 1         | 2        | 0.039776     | 3469012.450686     | 154580.940736 |

+-----------+----------+--------------+--------------------+---------------+

SUCCESS: Optimal solution found.

In [19]:
## compare the two models

In [20]:
my_feature_model.evaluate(test_data)

{'max_error': 3486584.509381705, 'rmse': 179542.4333126903}

In [21]:
advanced_feature_model.evaluate(test_data)

{'max_error': 3556849.413858208, 'rmse': 156831.1168021901}

# Question 3: my_feature_model: rmse=179542.43, advanced_feature_model=156831.11